File: src/doob/strategies/intraday_drift.py
"""Intraday Drift Backtesting Engine.

Buy SPY at the open, sell at the close on the same day.
This captures the intraday return component, complementary to the overnight drift.

Note: adj_close == close in this warehouse (IB TRADES data, split-adjusted
but not dividend-adjusted). For intraday returns ln(Close_t / Open_t)
this is fine — both prices are same-day so no dividend adjustment needed.
"""

from __future__ import annotations

import argparse
import logging
from pathlib import Path

import matplotlib
import numpy as np
import pandas as pd

matplotlib.use("Agg")
import matplotlib.pyplot as plt  # noqa: E402

from doob.config import output_dir
from doob.data.readers import load_ticker_ohlcv
from doob.metrics.fees import ibkr_roundtrip_cost
from doob.metrics.performance import (
    annual_returns_table,
    cagr,
    max_drawdown,
    sharpe,
    var_95,
)

log = logging.getLogger(__name__)

DEFAULT_CAPITAL = 1_000_000.0
OUTPUT_DIR = output_dir()


# ---------------------------------------------------------------------------
# Signal computation
# ---------------------------------------------------------------------------
def compute_intraday_returns(df: pd.DataFrame) -> pd.Series:
    """Compute intraday log returns: ln(Close_t / Open_t)."""
    return np.log(df["close"] / df["open"])


# ---------------------------------------------------------------------------
# Strategy simulation
# ---------------------------------------------------------------------------
def simulate_strategy(
    opens: np.ndarray,
    closes: np.ndarray,
    mask: np.ndarray,
    capital: float = DEFAULT_CAPITAL,
    fee_fn=ibkr_roundtrip_cost,
    short: bool = False,
) -> np.ndarray:
    """Simulate intraday strategy with equity-tracking loop.

    Long: buy at open, sell at close same day.
    Short: sell at open, cover at close same day.

    Returns:
        equity curve array (length = len(opens) + 1, starting at capital)
    """
    n = len(opens)
    equity = np.empty(n + 1)
    equity[0] = capital
    current = capital
    direction = -1.0 if short else 1.0

    for i in range(n):
        if mask[i]:
            shares = int(current / opens[i])
            if shares > 0:
                cost = fee_fn(current, opens[i])
                pnl = direction * shares * (closes[i] - opens[i])
                current = current + pnl - cost
        equity[i + 1] = current

    return equity


# ---------------------------------------------------------------------------
# Plotting
# ---------------------------------------------------------------------------
def plot_dashboard(  # pragma: no cover
    results: dict,
    dates: pd.Series,
    ticker: str = "SPY",
    output_dir: Path = OUTPUT_DIR,
) -> None:
    """Render equity curves and rolling volatility charts."""
    output_dir.mkdir(parents=True, exist_ok=True)
    slug = ticker.lower()

    fig, ax = plt.subplots(figsize=(14, 7))
    for name, data in results.items():
        ax.plot(dates, data["equity"][1:], label=name, linewidth=0.8)
    ax.set_title(f"Buy the Open, Sell the Close ({ticker}) — Equity Curves")
    ax.set_ylabel("Portfolio Value ($)")
    ax.set_xlabel("Date")
    ax.legend()
    ax.grid(True, alpha=0.3)
    ax.set_yscale("log")
    fig.tight_layout()
    fig.savefig(output_dir / f"intraday_drift_{slug}_equity.png", dpi=150)
    plt.close(fig)

    fig, ax = plt.subplots(figsize=(14, 5))
    for name, data in results.items():
        daily_ret = np.diff(data["equity"]) / data["equity"][:-1]
        rolling_vol = pd.Series(daily_ret).rolling(63).std() * np.sqrt(252)
        ax.plot(dates, rolling_vol.values, label=name, linewidth=0.8)
    ax.set_title(f"Buy the Open, Sell the Close ({ticker}) — Rolling 63-day Annualized Volatility")
    ax.set_ylabel("Volatility")
    ax.set_xlabel("Date")
    ax.legend()
    ax.grid(True, alpha=0.3)
    fig.tight_layout()
    fig.savefig(output_dir / f"intraday_drift_{slug}_volatility.png", dpi=150)
    plt.close(fig)

    log.info("Charts saved to %s", output_dir)


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------
def main() -> None:  # pragma: no cover
    parser = argparse.ArgumentParser(description="Buy the Open, Sell the Close Backtest")
    parser.add_argument("--start-date", type=str, default=None, help="Start date (YYYY-MM-DD)")
    parser.add_argument("--end-date", type=str, default=None, help="End date (YYYY-MM-DD)")
    parser.add_argument("--capital", type=float, default=DEFAULT_CAPITAL, help="Starting capital")
    parser.add_argument("--no-plots", action="store_true", help="Skip chart generation")
    parser.add_argument("--ticker", type=str, default="SPY", help="Ticker symbol (default: SPY)")
    parser.add_argument("--short", action="store_true", help="Short at open, cover at close")
    parser.add_argument("--start-year-table", type=int, default=2015, help="Annual table start year")
    args = parser.parse_args()

    ticker = args.ticker.upper()
    logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")

    print(f"Loading {ticker} from bronze parquet...")
    spy = load_ticker_ohlcv(ticker)
    print(f"  {ticker}: {len(spy):,} bars, {spy['trade_date'].min().date()} to {spy['trade_date'].max().date()}")

    if args.start_date:
        spy = spy[spy["trade_date"] >= pd.Timestamp(args.start_date)]
    if args.end_date:
        spy = spy[spy["trade_date"] <= pd.Timestamp(args.end_date)]
    spy = spy.reset_index(drop=True)

    dates = spy["trade_date"]
    n = len(spy)
    opens = spy["open"].values
    closes = spy["close"].values

    strategies = {}

    bh_equity = np.empty(n + 1)
    bh_equity[0] = args.capital
    shares_bh = int(args.capital / closes[0])
    cost_bh = ibkr_roundtrip_cost(args.capital, closes[0])
    for i in range(n):
        bh_equity[i + 1] = shares_bh * closes[i] + (args.capital - shares_bh * closes[0]) - cost_bh
    strategies["Buy & Hold"] = {"equity": bh_equity}

    mask_all = np.ones(n, dtype=bool)
    eq_intra = simulate_strategy(opens, closes, mask_all, args.capital, short=args.short)
    label = "Short Open→Cover Close" if args.short else "Intraday (Open→Close)"
    strategies[label] = {"equity": eq_intra}

    years = (dates.iloc[-1] - dates.iloc[0]).days / 365.25

    print("\n" + "=" * 80)
    mode = "SHORT OPEN, COVER CLOSE" if args.short else "BUY THE OPEN, SELL THE CLOSE"
    print(f"{mode} — {ticker} BACKTEST RESULTS")
    print(f"Period: {dates.iloc[0].date()} to {dates.iloc[-1].date()} ({years:.1f} years)")
    print(f"Capital: ${args.capital:,.0f} | Fee model: IBKR Tiered")
    print("Note: adj_close == close (IB split-adj only); B&H CAGR understates by ~1.3%/yr")
    print("=" * 80)

    header = f"{'Strategy':<25} {'Final ($)':>14} {'CAGR':>8} {'Sharpe':>8} {'MaxDD':>8} {'VaR95':>8}"
    print(header)
    print("-" * len(header))

    for name, data in strategies.items():
        eq = data["equity"]
        daily_returns = np.diff(eq) / eq[:-1]
        c = cagr(eq, years)
        s = sharpe(daily_returns)
        md = max_drawdown(eq)
        v = var_95(daily_returns)
        print(f"{name:<25} {eq[-1]:>14,.0f} {c:>7.1%} {s:>8.2f} {md:>7.1%} {v:>8.4f}")

    print(f"\nAnnual Returns (from {args.start_year_table}):")
    ann_header = f"{'Year':<6}"
    for name in strategies:
        ann_header += f" {name:>25}"
    print(ann_header)
    print("-" * len(ann_header))

    annual_tables = {}
    for name, data in strategies.items():
        annual_tables[name] = annual_returns_table(data["equity"], dates, args.start_year_table)

    all_years: set[int] = set()
    for tbl in annual_tables.values():
        all_years.update(tbl["year"].tolist())

    for year in sorted(all_years):
        row = f"{year:<6}"
        for name in strategies:
            tbl = annual_tables[name]
            yr_row = tbl[tbl["year"] == year]
            if len(yr_row) > 0:
                row += f" {yr_row['return'].iloc[0]:>24.1%}"
            else:
                row += f" {'N/A':>25}"
        print(row)

    if not args.no_plots:
        print(f"\nGenerating charts to {OUTPUT_DIR}/...")
        plot_dashboard(results=strategies, dates=dates, ticker=ticker)
        print("Done.")


if __name__ == "__main__":
    main()

File: src/strategies/intraday_drift.rs
/// Intraday Drift Backtesting Engine.
///
/// Buy at the open, sell at the close same day. Optional short mode.

use anyhow::Result;
use chrono::NaiveDate;

use num_format::ToFormattedString;

use crate::cli::OutputFormat;
use crate::data::readers::{load_ticker_ohlcv, OhlcvRow};
use crate::metrics::fees::ibkr_roundtrip_cost;
use crate::strategies::common::{
    build_json_annual_returns, buy_and_hold_equity, compute_strategy_metrics, format_annual_table,
    format_results_header, format_strategy_row, JsonOutput, StrategyResult,
};

pub const DEFAULT_CAPITAL: f64 = 1_000_000.0;

/// Compute intraday log returns: ln(Close_t / Open_t).
pub fn compute_intraday_returns(rows: &[OhlcvRow]) -> Vec<f64> {
    rows.iter().map(|r| (r.close / r.open).ln()).collect()
}

/// Simulate intraday strategy with equity-tracking loop.
///
/// Long: buy at open, sell at close same day.
/// Short: sell at open, cover at close same day.
pub fn simulate_strategy(
    opens: &[f64],
    closes: &[f64],
    mask: &[bool],
    capital: f64,
    fee_fn: &dyn Fn(f64, f64) -> f64,
    short: bool,
) -> Vec<f64> {
    let n = opens.len();
    let mut equity = Vec::with_capacity(n + 1);
    equity.push(capital);
    let mut current = capital;
    let direction: f64 = if short { -1.0 } else { 1.0 };

    for i in 0..n {
        if mask[i] {
            let shares = (current / opens[i]) as i64;
            if shares > 0 {
                let cost = fee_fn(current, opens[i]);
                let pnl = direction * shares as f64 * (closes[i] - opens[i]);
                current = current + pnl - cost;
            }
        }
        equity.push(current);
    }

    equity
}

/// CLI arguments for the intraday-drift strategy.
#[derive(Debug, clap::Args)]
pub struct IntradayDriftArgs {
    #[arg(long, help = "Start date (YYYY-MM-DD)")]
    pub start_date: Option<String>,

    #[arg(long, help = "End date (YYYY-MM-DD)")]
    pub end_date: Option<String>,

    #[arg(long, default_value_t = DEFAULT_CAPITAL, help = "Starting capital")]
    pub capital: f64,

    #[arg(long, help = "Skip chart generation")]
    pub no_plots: bool,

    #[arg(long, default_value = "SPY", help = "Ticker symbol")]
    pub ticker: String,

    #[arg(long, help = "Short at open, cover at close")]
    pub short: bool,

    #[arg(long, default_value_t = 2015, help = "Annual table start year")]
    pub start_year_table: i32,
}

/// Run the intraday drift strategy.
pub fn run(args: &IntradayDriftArgs, fmt: OutputFormat) -> Result<()> {
    let json_mode = fmt == OutputFormat::Json;
    let ticker = args.ticker.to_uppercase();

    if !json_mode {
        println!("Loading {} from bronze parquet...", ticker);
    }
    let mut spy = load_ticker_ohlcv(&ticker, None, None)?;
    if !json_mode {
        println!(
            "  {}: {} bars, {} to {}",
            ticker,
            spy.len(),
            spy.first().unwrap().trade_date,
            spy.last().unwrap().trade_date
        );
    }

    // Date filtering
    if let Some(ref sd) = args.start_date {
        let start = sd.parse::<NaiveDate>()?;
        spy.retain(|r| r.trade_date >= start);
    }
    if let Some(ref ed) = args.end_date {
        let end = ed.parse::<NaiveDate>()?;
        spy.retain(|r| r.trade_date <= end);
    }

    let n = spy.len();
    let dates: Vec<NaiveDate> = spy.iter().map(|r| r.trade_date).collect();
    let opens: Vec<f64> = spy.iter().map(|r| r.open).collect();
    let closes: Vec<f64> = spy.iter().map(|r| r.close).collect();

    let mut strategies: Vec<StrategyResult> = Vec::new();

    // Buy & Hold
    let bh_equity = buy_and_hold_equity(&closes, args.capital);
    strategies.push(StrategyResult {
        name: "Buy & Hold".to_string(),
        equity: bh_equity,
    });

    // Intraday
    let mask_all = vec![true; n];
    let eq_intra = simulate_strategy(
        &opens,
        &closes,
        &mask_all,
        args.capital,
        &ibkr_roundtrip_cost,
        args.short,
    );
    let label = if args.short {
        "Short Open→Cover Close"
    } else {
        "Intraday (Open→Close)"
    };
    strategies.push(StrategyResult {
        name: label.to_string(),
        equity: eq_intra,
    });

    let years = (dates.last().unwrap().signed_duration_since(*dates.first().unwrap())).num_days()
        as f64
        / 365.25;

    if json_mode {
        let results: Vec<_> = strategies
            .iter()
            .map(|s| compute_strategy_metrics(&s.name, &s.equity, years))
            .collect();

        let output = JsonOutput {
            strategy: "intraday-drift".to_string(),
            ticker: ticker.clone(),
            period_start: dates.first().unwrap().to_string(),
            period_end: dates.last().unwrap().to_string(),
            years,
            capital: args.capital,
            fee_model: "IBKR Tiered".to_string(),
            results,
            annual_returns: build_json_annual_returns(
                &strategies,
                &dates,
                args.start_year_table,
            ),
        };
        println!("{}", serde_json::to_string(&output)?);
    } else {
        println!();
        println!("{}", "=".repeat(80));
        let mode = if args.short {
            "SHORT OPEN, COVER CLOSE"
        } else {
            "BUY THE OPEN, SELL THE CLOSE"
        };
        println!("{} — {} BACKTEST RESULTS", mode, ticker);
        println!(
            "Period: {} to {} ({:.1} years)",
            dates.first().unwrap(),
            dates.last().unwrap(),
            years
        );
        println!(
            "Capital: ${} | Fee model: IBKR Tiered",
            (args.capital as i64).to_formatted_string(&num_format::Locale::en)
        );
        println!("Note: adj_close == close (IB split-adj only); B&H CAGR understates by ~1.3%/yr");
        println!("{}", "=".repeat(80));

        let (header, sep) = format_results_header();
        println!("{header}");
        println!("{sep}");

        for strat in &strategies {
            println!("{}", format_strategy_row(&strat.name, &strat.equity, years));
        }

        println!("\nAnnual Returns (from {}):", args.start_year_table);
        println!(
            "{}",
            format_annual_table(&strategies, &dates, args.start_year_table, 25)
        );

        if !args.no_plots {
            println!("\nPlotting not yet implemented in Rust port (use --no-plots)");
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::data::readers::OhlcvRow;

    fn make_rows(opens: &[f64], closes: &[f64]) -> Vec<OhlcvRow> {
        opens
            .iter()
            .zip(closes.iter())
            .enumerate()
            .map(|(i, (&o, &c))| OhlcvRow {
                trade_date: NaiveDate::from_ymd_opt(2024, 1, 2 + i as u32).unwrap(),
                open: o,
                high: o.max(c) + 1.0,
                low: o.min(c) - 1.0,
                close: c,
                volume: 1_000_000.0,
            })
            .collect()
    }

    #[test]
    fn test_compute_intraday_returns() {
        let rows = make_rows(&[100.0, 102.0, 104.0], &[101.0, 103.0, 105.0]);
        let ret = compute_intraday_returns(&rows);
        assert!((ret[0] - (101.0_f64 / 100.0).ln()).abs() < 1e-10);
        assert!((ret[1] - (103.0_f64 / 102.0).ln()).abs() < 1e-10);
        assert!((ret[2] - (105.0_f64 / 104.0).ln()).abs() < 1e-10);
    }

    #[test]
    fn test_simulate_basic() {
        let opens = [100.0, 101.0, 102.0];
        let closes = [101.0, 102.0, 103.0];
        let mask = [true, true, true];

        let equity = simulate_strategy(&opens, &closes, &mask, 10_000.0, &|_, _| 0.0, false);
        assert_eq!(equity.len(), 4);
        assert_eq!(equity[0], 10_000.0);
        assert!((equity[1] - 10_100.0).abs() < 1e-6);
    }

    #[test]
    fn test_mask_skips_trades() {
        let opens = [100.0, 100.0];
        let closes = [110.0, 110.0];
        let mask = [false, true];

        let equity = simulate_strategy(&opens, &closes, &mask, 10_000.0, &|_, _| 0.0, false);
        assert_eq!(equity[1], 10_000.0);
        assert!(equity[2] > 10_000.0);
    }

    #[test]
    fn test_loss_day() {
        let opens = [100.0];
        let closes = [95.0];
        let mask = [true];

        let equity = simulate_strategy(&opens, &closes, &mask, 10_000.0, &|_, _| 0.0, false);
        assert!((equity[1] - 9_500.0).abs() < 1e-6);
    }

    #[test]
    fn test_fees_reduce_equity() {
        let opens = [100.0];
        let closes = [100.0];
        let mask = [true];

        let equity =
            simulate_strategy(&opens, &closes, &mask, 10_000.0, &ibkr_roundtrip_cost, false);
        assert!(equity[1] < 10_000.0);
    }

    #[test]
    fn test_short_direction() {
        let opens = [100.0];
        let closes = [95.0];
        let mask = [true];

        let equity = simulate_strategy(&opens, &closes, &mask, 10_000.0, &|_, _| 0.0, true);
        // Short: pnl = -1 * 100 * (95-100) = 500
        assert!((equity[1] - 10_500.0).abs() < 1e-6);
    }

    #[test]
    fn test_negative_return() {
        let rows = make_rows(&[105.0], &[100.0]);
        let ret = compute_intraday_returns(&rows);
        assert!(ret[0] < 0.0);
    }

    #[test]
    fn test_flat_day() {
        let rows = make_rows(&[100.0], &[100.0]);
        let ret = compute_intraday_returns(&rows);
        assert!((ret[0]).abs() < 1e-10);
    }
}
